Tags

davidhalter opened the pull request:
Fixes #35

davidhalter committed:
Implement tags for volumes

davidhalter committed:
Add server tags

davidhalter committed:
Test that the number of tags actually matches tests to terraform state

davidhalter committed:
Test for servers as well that tags are part of Captain

davidhalter committed:
Implement Floating IP tags

davidhalter committed:
Remove tags from gathering of volume resource data

davidhalter committed:
Implement Server Group tags

davidhalter committed:
Allow network tags

davidhalter committed:
Implement subnet tags

davidhalter committed:
Implement custom image tags

davidhalter committed:
Implement objects user tags

davidhalter committed:
Documentation about tags

davidhalter committed:
Fix a small documentation issue around floating IP datasources

davidhalter committed:
Fix server tags

alakae commented on vendor/github.com/cloudscale-ch/cloudscale-go-sdk/tags.go:
```diff
@@ -12,7 +12,7 @@ type TaggedResource struct {
 
 type TaggedResourceRequest struct {
-	Tags TagMap `json:"tags,omitempty"`
+	Tags *TagMap `json:"tags,omitempty"`
```
Please run a proper `go get -u` and `go mod vendor` after releasing the new go-sdk.

davidhalter replied:
Why do a `go get -u`? This causes a lot of indirect updates and will update pretty much everything? I will of course vendor the new cloudscale-go-sdk.

alakae replied:
Excuse me, I was just sloppy when writing this comment. I meant `go get -u 'github.com/cloudscale-ch/cloudscale-go-sdk'`

davidhalter replied:
It's `go get -u 'github.com/cloudscale-ch/cloudscale-go-sdk/v2'` now... *sigh*...

Golang :shrug:

alakae commented on docs/resources/subnet.md:
```diff
@@ -54,6 +54,13 @@ The following arguments are supported when creating/changing subnets:
 * `network_uuid` - (Required) The network of the subnet.
 * `gateway_address` - (Optional) The gateway address of the subnet.
 * `dns_servers` - (Optional) A list of DNS resolver IP addresses, that act as DNS servers.
+* `tags` - (Optional) Tags can be used to specify additional information on resources:
```
I suggest that we use the same wording as in our API docs:
> Tags allow you to assign custom metadata to resources

alakae commented on cloudscale/resource_cloudscale_server_group.go:
```diff
@@ -46,6 +46,7 @@ func getServerGroupSchema(t SchemaType) map[string]*schema.Schema {
 			Type:     schema.TypeString,
 			Computed: true,
 		},
+		"tags": &TagsSchema,
```
There is a problem with importing tags. Suppose there is a server group:

```json
{
    "href": "https://api.cloudscale.ch/v1/server-groups/db75d354-07dd-4eca-9484-07157280aa92",
    "uuid": "db75d354-07dd-4eca-9484-07157280aa92",
    "name": "g2",
    "servers": [],
    "type": "anti-affinity",
    "zone": {
      "slug": "lpg1"
    },
    "tags": {
      "status": "completed",
      "project": "gemini"
    }
  }
```

and a resource configuration:

```
resource "cloudscale_server_group" "g2" {
  name      = "g2"
  type      = "anti-affinity"
  zone_slug = "lpg1"
}
```

and we import it: 
```bash
$  terraform import  "cloudscale_server_group.g2" db75d354-07dd-4eca-9484-07157280aa92
cloudscale_server_group.g2: Importing from ID "db75d354-07dd-4eca-9484-07157280aa92"...
cloudscale_server_group.g2: Import prepared!
  Prepared cloudscale_server_group for import
cloudscale_server_group.g2: Refreshing state... [id=db75d354-07dd-4eca-9484-07157280aa92]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
```

Then the plan is empty:
```bash
$ terraform plan
cloudscale_server_group.g2: Refreshing state... [id=db75d354-07dd-4eca-9484-07157280aa92]

No changes. Your infrastructure matches the configuration.
```

Expected behaviour is that `terraform plan` detects the missing tags in tf configuration.

davidhalter replied:
This should be working now. Please check again.

alakae commented:
If possible, please make it possible to lookup tags on datasources.

davidhalter committed:
Tags documentation change

davidhalter committed:
Fix tags state after importing

alakae committed:
Merge remote-tracking branch 'upstream/master' into dave/tags

alakae commented on cloudscale/resource_cloudscale_server_group_test.go:
```diff
@@ -183,6 +183,53 @@ func TestAccCloudscaleServerGroup_import_basic(t *testing.T) {
 	})
 }
 
+func TestAccCloudscaleServerGroup_import_withTags(t *testing.T) {
```
Maybe add this kind of test case for all resources?

davidhalter replied:
It was intentionally not added for all resources. I feel like this is a bit much. I'm both trying to balance dev time and resource creation with the advantage of the tests and I feel like it's not necessary here. There are a ton of small things that are not well tested in Terraform, I feel like.

alakae commented on cloudscale/resource_cloudscale_server.go:
```diff
@@ -354,6 +354,7 @@ func resourceServerCreate(d *schema.ResourceData, meta interface{}) error {
 	if attr, ok := d.GetOk("status"); ok {
 		originalStatus = attr.(string)
 	}
+	opts.Tags = CopyTags(d)
```
I assume you found this issue yourself? It was not something that I reported?

davidhalter replied:
Yes, the tests were failing, because of this.

davidhalter committed:
Upgrade cloudscale-go-sdk to v2

Did this:
go get -u 'github.com/cloudscale-ch/cloudscale-go-sdk/v2@v2.0.1
go mod vendor

davidhalter committed:
Merge branch 'dave/tags' of ssh://github.com/cloudscale-ch/terraform-provider-cloudscale into dave/tags

davidhalter committed:
Properly use cloudscale-go-sdk/v2

davidhalter committed:
Vendor the new v2 packages
